File: Arrays/image-overlap.py
# 835. Image Overlap
# https://leetcode.com/problems/image-overlap/description/

import numpy as np
class Solution(object):
    def largestOverlap(self, img1, img2):
        """
        :type img1: List[List[int]]
        :type img2: List[List[int]]
        :rtype: int
        """
        a = np.array(img1)
        b = np.array(img2)
        n = a.shape[0]
        a_pad = np.pad(a, n-1, mode='constant')
        new_dim = 2 * n - 1
        conv_matrix = np.zeros(shape=(new_dim, new_dim))
        for r in range(new_dim):
            for c in range(new_dim):
                conv_matrix[r, c] = np.sum(b * a_pad[r:r+n,c:c+n])
        return int(np.max(conv_matrix))

File: Arrays/image-overlap.cs
// 835. Image Overlap
// https://leetcode.com/problems/image-overlap/description/

public class Solution {
    public int LargestOverlap(int[][] img1, int[][] img2) {
        int maxOverlap = int.MinValue;
        int n = img1.Length;
        for(int r=-n+1; r < n; r++){
            for(int c=-n+1; c < n ; c++){
               maxOverlap = Math.Max(GetOnesCount(img1, img2, r, c), maxOverlap);
            }
        }
        return maxOverlap;
    }

    private int GetOnesCount(int[][] img1, int[][] img2, int rowOffset, int colOffset) {
        int onesCount = 0;
        int n = img1.Length;
        for(int r=0; r < n; r++){
            for(int c=0; c < n ; c++){
                if(IsInBound(n, r+rowOffset, c+colOffset) && img2[r][c] + img1[r+rowOffset][c+colOffset] == 2)
                    onesCount++;
            }
        }
        return onesCount;
    }

    private bool IsInBound(int n, int row, int col){
        if(row < 0 || row >= n || col < 0 || col >= n) return false;
        return true;
    }

}